### PR TITLE
Download dataset

### PR DESCRIPTION
**Pull request by blue442**
Addressing issue #400 - this PR builds in the functionality required to download and access a dataset using the foundry sdk.  
Note: github doesn't recognize that the `dev` and `new_search` branches had a common merge in the past and the diff reverts to their last known common commit (I think this is because I continued comitting to a branch that was deleted in the repo instead of creating a new feature branch). Viewing the diff between the latest state of both branches (two-dot comparison) better visualizes the contributions of the latest commit, but it appears they can't be automatically merged.

Note that the issue (#400) was modified to reflect the data being loaded as a dictionary vs a pandas dataframe in order to recreate the functionality as it was prior to the refactoring. This allowed it to better serve as a drop-in replacement with the older implementation.

### Specific changes to the main code include:

- `Foundry.py`
   - add a `HiddenColumnDataFrame` object to the `foundry.py` module to mimic earlier display of search results while providing access to associated `FoundryDataset` objects for each entry (via the `get_dataset_by_[name/doi]()` function
   - add an `as_list = False` argument to the `search()` function to control whether the results are returned as a list or `HiddenColumnDataFrame`
   - Altered search implementation so that when a query string is provided, the behavior is to pull all foundry results from MDF and filter them locally by flattening the metadata for each dataset and looking for a case-insensitive match to the query string. This provides a more comprehensive search than the limited options provided by the `forge_client`, but may not be sustainable as the number of entries grows.

- `foundry_cache.py`
   - shifts authentication details to be passed into the Cache
   - implements `load_as_dict()` function as well as supporting functions to allow for loading the data into memory in the same fashion as in the previous iteration of `Foundry`

- `foundry_dataset.py`
   - implements functionality to download dataset to local storage
   - implements `get_as_dict()` function to load dataset into memory (serves as a wrapper for the similarly named function from the `FoundryCache` class)
   - implements functionality to remove locally stored datasets

### Global improvements include:

- removing of unused code
- establishment or improvement of docstrings for all functions
- shuffling a few things around to avoid circular imports

This PR also updates all of the examples to illustrate the new use case. This also includes removal of model running via foundry, as dl_hub is to be deprecated in the near future. Because of this, some of the examples that only focused on performing inference using a dl_hub model (and using something like `wget` for data access) were removed entirely.

### Testing updates
- included additional testing to verify the loading of a dataset as a dictionary is functioning properly
- tested `search()` returning results as a list
- modified testing of `search()` with query
- specific test to ensure metadata is pulled correctly
- test to ensure download via globus is working properly
- removed test of download via http, as the test was non-functional (verified that the function downloaded something, but that something was an html authorization page instead of a dataset)
- globally adjusted syntax of tests so the use of foundry matches the latest implementation

### CODE REVIEW

**Review by BenGalewsky — CHANGES_REQUESTED** [review]
This is looking good. Just a few suggested changes to the interface and the boundaries between the classes. 

Also there should be unit tests for `Dataset` and for `Cache` separately - those classes should then be mocked in `test_foundry`

File: foundry/foundry_dataset.py [hunk 1]
```diff
@@ -28,10 +39,54 @@ class FoundryDataset(FoundrySchema):
...
         """
 
-    def __init__(self, source_id: str, datacite_entry: dict, foundry_schema: FoundrySchema):
-        self.source_id = source_id
+    def __init__(self,
+                 dataset_name: str,
+                 datacite_entry: dict,
+                 transfer_client: Any,
+                 foundry_schema: FoundrySchema,
+                 globus: bool = False,
```
[the remark this answers is not in the document]
Comment (blue442) [reply foundry/foundry_dataset.py:47]: `globus` arg has been changed to `use_globus` (including in the examples).

File: foundry/foundry_dataset.py [hunk 2]
```diff
@@ -28,10 +39,54 @@ class FoundryDataset(FoundrySchema):
...
+        self.transfer_client = transfer_client
         self.foundry_schema = foundry_schema
+        self.globus = globus
+        self.interval = interval
+        self.parallel_https = parallel_https
+        self.verbose = verbose
+        self._foundry_cache = FoundryCache(forge_client, transfer_client)
+
+    def get_as_dict(self, split: str = None):
+        """Convert FoundryDatset to a Pandas Dataframe object
```
[the remark this answers is not in the document]
Comment (blue442) [reply foundry/foundry_dataset.py:64]: Yup - changed to dict

File: foundry/foundry_dataset.py [hunk 3]
```diff
@@ -28,10 +39,54 @@ class FoundryDataset(FoundrySchema):
...
+
+        Arguments:
+            split (string): Split to create PyTorch Dataset on.
+                    **Default:** ``None``
+
+        Returns: (dict) Dictionary of all the data from the specified split
+
+        """
+        self.download_if_not_downloaded()
+        return self._foundry_cache.load_as_dict(self.dataset_name,
```
[the remark this answers is not in the document]
Comment (blue442) [reply foundry/foundry_dataset.py:74]: done.

File: foundry/foundry_dataset.py [hunk 4]
```diff
@@ -67,55 +122,13 @@ def to_tensorflow(self, split: str = None):
...
-            else:
-                # raise an error if splits are specified but not present in the dataset
-                if len(splits) > 0:
-                    raise ValueError(f"Splits to load were specified as {splits}, but no splits are present in dataset")
-                return {"data": self._load_data(source_id=source_id, globus=globus, as_hdf5=as_hdf5)}
-        except Exception as e:
-            raise Exception(
-                "Metadata not loaded into Foundry object, make sure to call load()") from e
+        self._foundry_cache.download_to_cache(self.dataset_name,
+                                              self.foundry_schema.splits,
```
[the remark this answers is not in the document]
Comment (blue442) [reply foundry/foundry_dataset.py:126]: That was my initial instinct as well, but that creates a circular import.

File: foundry/foundry_cache.py [hunk 5]
```diff
@@ -2,97 +2,122 @@
...
 
 
 class FoundryCache():
     """The FoundryCache manages the local storage of FoundryDataset objects"""
 
-    def __init__(self):
+    def __init__(self,
+                 forge_client: Forge,
+                 transfer_client: Any):
+        """
```
[the remark this answers is not in the document]
Comment (blue442) [reply foundry/foundry_cache.py:25]: Added an optional arg to the `Foundry` constructor method - priority of locations for storing data is now

1. user specified `local_cache_dir` passed into `Foundry` init
2. env variable
3. './data'

Note that the argument to the `Foundry` constructor is not persisted between sessions.

**Comment by blue442** [comment]
Testing has been adjusted to be compatible with github actions:

- The globus-dependent test have been conditionally skipped in GHA using `@pytest.mark.skipif(bool(is_gha))`.
- The https download is now skipped to be pushed to a new issue (looks to be fixed in main, but needs to be incorporated). 
- It appears that GHA may be having issues with the presence of `pytest.raises()` assertions, causing the GHA testing to time out. These tests have been marked with `@pytest.mark.skipif(bool(is_gha))` so that they are skipped during github actions, but all run fine locally.

Re: @BenGalewsky 's comment, testing for foundry cache and dataset objects have been separated out.

**Review by ascourtas — COMMENTED** [review]
Okey doke, this was a big ol PR! This will cover *a lot* of major changes for the refactor and is a huge step forward in Foundry. Thank you @blue442 for all your hard work! 

I just had a few minor comments here. For the rest of the review, I just need to go through the notebooks (it doesnt work well in github), and have @kjschmidt913 give the final AOK to make sure that the user experience is what we want it to be :) 

Great work! 🎉

File: foundry/foundry.py [hunk 6]
```diff
@@ -1,38 +1,90 @@
...
 
 from foundry.https_upload import upload_to_endpoint
 
 logging.basicConfig(format='%(levelname)s: %(message)s', level=logging.WARNING)
 logger = logging.getLogger(__name__)
 
 
+class HiddenColumnDataFrame(pd.DataFrame):
+    """
+    A subclass of pd.DataFrame that supports hiding a specific column.
```
[the remark this answers is not in the document]
Comment (blue442) [reply foundry/foundry.py:32]: Added additional context to docstring!

File: foundry/foundry.py [hunk 7]
```diff
@@ -203,13 +297,14 @@ def dataset_from_metadata(self, metadata: dict) -> FoundryDataset:
     def get_dataset_by_name(self, name: str) -> FoundryDataset:
```
[the remark this answers is not in the document]
Comment (blue442) [reply foundry/foundry.py:297]: The `get_metadata_by_*()` functions are really just helper functions, and could be marked as private if we want to go that route. They are used fetch the metadata which is then converted to a `FoundryDataset` object before being returned to the user - so I think the specificity of the name is important. Issue #400 specified that passing a DOI to the `search()` method would work appropriately, so the helper method had to be introduced, but the user-facing (and similarly named) `get_dataset_by_doi()` isn't a deliverable until #402. 

The `get_dataset_by_name()` function looks to have been pre-emptively implemented, as it isn't part of the acceptance criteria of #400 and isn't referenced anywhere in tests or notebooks. I'm believe that it's meant to be part of the eventually functionality provided to the user (@kjschmidt913 can you verify?) but maybe we should leave it out until it's specifically requested?

File: foundry/foundry_dataset.py [hunk 8]
```diff
@@ -28,94 +36,98 @@ class FoundryDataset(FoundrySchema):
...
+            split (string): Split to create dataset on.
+                    **Default:** ``None``
 
-        from foundry.loaders.torch_wrapper import TorchDataset
+        Returns: (Pandas.Dataframe) Pandas ddataframe of all the data from the specified split
 
-        inputs, targets = self._get_inputs_targets(split)
-        return TorchDataset(inputs, targets)
+        """
+        pass
```
[the remark this answers is not in the document]
Comment (blue442) [reply foundry/foundry_dataset.py:91]: The `_repr_html_()` override was done in the `foundry.HiddenColmnDataFrame` class - this one is a residual from the original assumption when speccing out the issue that datasets were originally returned as `Pandas.Dataframe` objects (vs dictionaries). I'll remove for now as it isn't part of the acceptance criteria. If we want to pursue it we can flesh out what a dataset would look like in a `Pandas.Dataframe` format when discussion the acceptance criteria for the stubbed issue (#428).

File: tests/test_foundry_components.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    elwood_data = search_results[0].get_as_dict()
+    X, y = elwood_data['train']
+
+    assert len(X) > 1
+    assert isinstance(X, pd.DataFrame)
+    assert len(y) > 1
+    assert isinstance(y, pd.DataFrame)
+
+
+# def test_download_via_globus(auths, elwood_data, mocker):
```
[the remark this answers is not in the document]
Comment (blue442) [reply tests/test_foundry_components.py:88]: That was fully extraneous - removed.

File: foundry/foundry_cache.py [hunk 10]
```diff
@@ -2,97 +2,122 @@
...
-        """Downloads the data from source to local storage
-
-        Args:
-            dataset: a FoundryDataset object created from the metadata
-            globus: if True, use Globus to download the data else try HTTPS
-            interval: How often to wait before checking Globus transfer status
-            parallel_https: Number of files to download in parallel if using HTTPS
-            verbose: Produce more debug messages to screen
+    def download_to_cache(self,
+                          dataset_name: str,
```
[the remark this answers is not in the document]
Comment (blue442) [reply foundry/foundry_cache.py:45]: Can confirm - avoiding circular imports is why I’m passing the individual attributes. I haven’t figured out any way around that.

File: foundry/foundry.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kjschmidt913) [reply foundry/foundry.py:297]: I think get rid of get_dataset_by_name()

[the remark this answers is not in the document]
Comment (blue442) [reply foundry/foundry.py:297]: Dropping it!

**Pull request merged**